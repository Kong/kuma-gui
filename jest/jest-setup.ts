import '@testing-library/jest-dom/extend-expect'
import '@testing-library/jest-dom'

// Polyfills `window.fetch` for Jest because it runs in a Node environment where fetch isn’t available. It initially looked like this would change with Node.js 18, but that is not so.
import 'isomorphic-fetch'

import dotenv from 'dotenv'

import { replaceAttributesSnapshotSerializer } from './jest-replace-attribute-snapshot-serializer'
import { setupMockServer } from '../src/api/setupMockServer'

/**
 * Ensures the tests run with the project’s environment variables being available.
 */
dotenv.config()

/**
 * amcharts need SVGPathElement defined for our tests to work.
 *
 * See: https://github.com/amcharts/amcharts4/issues/1387
 */
Object.defineProperty(window, 'SVGPathElement', { value: class extends HTMLElement { } })

/**
 * Kongponents uses generated UUIDs for several attribute values.
 * This breaks the project’s snapshot tests since they’re based on fully-mounted components
 * which also includes those from external sources like Kongponents.
 *
 * In order to stabilize the tests which otherwise fail because the attribute values are different every run,
 * we use a custom snapshot serializer to replace those attribute values with one fixed value.
 */
expect.addSnapshotSerializer(replaceAttributesSnapshotSerializer([
  'id',
  'aria-describedby',
  'aria-labelledby',
  'aria-controls',
  'data-tableid',
]))

const server = setupMockServer(import.meta.env.VITE_KUMA_API_SERVER_URL)

// Establish API mocking before all tests.
beforeAll(() => server.listen())
// Reset any request handlers that we may add during the tests,
// so they don't affect other tests.
afterEach(() => server.resetHandlers())
// Clean up after the tests are finished.
afterAll(() => server.close())

export { server }
