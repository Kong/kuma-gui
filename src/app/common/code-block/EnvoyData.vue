<template>
  <div>
    <DataSource
      v-slot="{ data, error, refresh }: EnvoyDataSource"
      :src="props.src"
    >
      <ErrorBlock
        v-if="error"
        :error="error"
      />

      <LoadingBlock v-else-if="data === undefined" />

      <EmptyBlock v-else-if="data === ''" />

      <CodeBlock
        v-else
        language="json"
        :code="typeof data === 'string' ? data : JSON.stringify(data, null, 2)"
        is-searchable
        :query="props.query"
        :is-filter-mode="props.isFilterMode"
        :is-reg-exp-mode="props.isRegExpMode"
        @query-change="emit('query-change', $event)"
        @filter-mode-change="emit('filter-mode-change', $event)"
        @reg-exp-mode-change="emit('reg-exp-mode-change', $event)"
      >
        <template #primary-actions>
          <KButton
            appearance="primary"
            @click="refresh"
          >
            <RefreshIcon />

            Refresh
          </KButton>
        </template>
      </CodeBlock>
    </DataSource>
  </div>
</template>

<script lang="ts" setup>
import { RefreshIcon } from '@kong/icons'

import CodeBlock from './CodeBlock.vue'
import EmptyBlock from '@/app/common/EmptyBlock.vue'
import ErrorBlock from '@/app/common/ErrorBlock.vue'
import LoadingBlock from '@/app/common/LoadingBlock.vue'
import { EnvoyDataSource } from '@/app/zones/sources'

const props = withDefaults(defineProps<{
  resource: string
  src: string
  query?: string
  isFilterMode?: boolean
  isRegExpMode?: boolean
}>(), {
  query: '',
  isFilterMode: false,
  isRegExpMode: false,
})

const emit = defineEmits<{
  (event: 'query-change', query: string): void
  (event: 'filter-mode-change', isFilterMode: boolean): void
  (event: 'reg-exp-mode-change', isRegExpMode: boolean): void
}>()
</script>
