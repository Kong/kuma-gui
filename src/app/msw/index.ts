import { setupWorker } from 'msw/browser'

import { token } from '@/services/utils'
import type { Token, ServiceDefinition } from '@/services/utils'
import type { SetupWorker } from 'msw/browser'

type Handler = Parameters<typeof setupWorker>[0]
type Passthrough = (req: Request) => string | string[] | void

type Msw = ReturnType<SetupWorker['start']>

const $ = {
  msw: token<Msw>('msw'),
  mswHandlers: token('msw.handlers'),
  mswPassthroughs: token('msw.passthroughs'),
}
export const TOKENS = $
const passthroughExtension = /\.(vue|ts|js|json)(\?.*)?$/
export const services = (_: Record<string, Token>): ServiceDefinition[] => {
  return [
    [token('msw.passthrough.msw'), {
      service: () => [(req: Request) => {
        const { pathname, href } = new URL(req.url)
        if (
          !pathname.startsWith('/node_modules') &&
         !pathname.startsWith('/src/assets/') &&
         !passthroughExtension.test(href)
        ) {
          return `Found an unhandled ${req.method} request to ${href}`
        }
      }],
      labels: [
        $.mswPassthroughs,
      ],
    },
    ],
    [$.msw, {
      service: (handlers: Handler[] = [], passthroughs: Passthrough[]) => {
        return setupWorker(...handlers).start({
          quiet: true,
          onUnhandledRequest(req: Request) {
            passthroughs.reduce<string[]>((prev, item) => prev.concat(item(req) ?? []), []).forEach(item => {
              console.warn(item)
            })
          },
        })
      },
      arguments: [
        $.mswHandlers,
        $.mswPassthroughs,
      ],
    }],
  ]
}
