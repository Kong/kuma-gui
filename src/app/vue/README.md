---
section: Packages
---
# vue

Automatic `vue` initialization using our projects dependency injection layer.

