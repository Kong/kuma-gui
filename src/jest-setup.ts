import '@testing-library/jest-dom/extend-expect'
import '@testing-library/jest-dom'
import { replaceAttributesSnapshotSerializer } from './jest-replace-attribute-snapshot-serializer'
import { server as setupServer } from '@/services/mocks'

import './kongponents'

class SVGPathElement extends HTMLElement { }

// @ts-ignore
window.SVGPathElement = SVGPathElement

/**
 * Kongponents v6 uses a call to `crypto.getRandomValues` which isn’t implemented in jsdom by default.
 * To make the tests run, we supply it via Node’s `crypto` library.
 *
 * See: https://github.com/jsdom/jsdom/issues/1612
 */
Object.defineProperty(global.self, 'crypto', {
  value: {
    getRandomValues: (array: []) => require('crypto').randomBytes(array.length),
  },
})

/**
 * Kongponents v6 uses generated UUIDs for several attribute values.
 * This breaks the project’s snapshot tests since they’re based on fully-mounted components
 * which also includes those from external sources like Kongponents.
 *
 * In order to stabilize the tests which otherwise fail because the attribute values are different every run,
 * we use a custom snapshot serializer to replace those attribute values with one fixed value.
 */
expect.addSnapshotSerializer(replaceAttributesSnapshotSerializer([
  'id',
  'aria-describedby',
  'aria-labelledby',
  'aria-controls',
]))

const server = setupServer('http://localhost/')

// Establish API mocking before all tests.
beforeAll(() => server.listen())
// Reset any request handlers that we may add during the tests,
// so they don't affect other tests.
afterEach(() => server.resetHandlers())
// Clean up after the tests are finished.
afterAll(() => server.close())

export { server }
