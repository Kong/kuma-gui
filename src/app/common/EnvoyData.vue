<template>
  <div class="envoy-data">
    <div class="envoy-data-actions">
      <KButton
        :disabled="isLoading"
        appearance="primary"
        icon="redo"
        data-testid="envoy-data-refresh-button"
        @click="fetchContent"
      >
        Refresh
      </KButton>
    </div>

    <StatusInfo
      :has-error="error !== null"
      :is-loading="isLoading"
      :error="error"
    >
      <CodeBlock
        :id="`code-block-${props.dataPath}`"
        language="json"
        :code="code"
        is-searchable
        :query-key="props.queryKey ?? `code-block-${props.dataPath}`"
      />
    </StatusInfo>
  </div>
</template>

<script lang="ts" setup>
import { onMounted, PropType, ref, watch } from 'vue'
import { KButton } from '@kong/kongponents'

import CodeBlock from './CodeBlock.vue'
import StatusInfo from './StatusInfo.vue'
import { useKumaApi } from '@/utilities'

const kumaApi = useKumaApi()

const props = defineProps({
  dataPath: {
    type: String as PropType<'xds' | 'stats' | 'clusters'>,
    required: true,
  },

  queryKey: {
    type: String,
    required: false,
    default: null,
  },

  mesh: {
    type: String,
    required: false,
    default: '',
  },

  dppName: {
    type: String,
    required: false,
    default: '',
  },

  zoneIngressName: {
    type: String,
    required: false,
    default: '',
  },

  zoneEgressName: {
    type: String,
    required: false,
    default: '',
  },
})

const isLoading = ref(true)
const error = ref<Error | null>(null)
const code = ref('')

watch(() => props.dppName, function () {
  fetchContent()
})
watch(() => props.zoneIngressName, function () {
  fetchContent()
})
watch(() => props.zoneEgressName, function () {
  fetchContent()
})

onMounted(function () {
  fetchContent()
})

async function fetchContent() {
  error.value = null
  isLoading.value = true

  try {
    let content = ''

    if (props.mesh !== '' && props.dppName !== '') {
      content = await kumaApi.getDataplaneData({
        dataPath: props.dataPath,
        mesh: props.mesh,
        dppName: props.dppName,
      })
    } else if (props.zoneIngressName !== '') {
      content = await kumaApi.getZoneIngressData({
        dataPath: props.dataPath,
        zoneIngressName: props.zoneIngressName,
      })
    } else if (props.zoneEgressName !== '') {
      content = await kumaApi.getZoneEgressData({
        dataPath: props.dataPath,
        zoneEgressName: props.zoneEgressName,
      })
    }

    code.value = typeof content === 'string' ? content : JSON.stringify(content, null, 2)
  } catch (err) {
    if (err instanceof Error) {
      error.value = err
    } else {
      console.error(err)
    }
  } finally {
    isLoading.value = false
  }
}
</script>

<style lang="scss" scoped>
.envoy-data {
  padding: var(--spacing-md);
}

.envoy-data-actions {
  display: flex;
  justify-content: flex-end;
  align-items: center;
  margin-bottom: var(--spacing-md);
}
</style>
