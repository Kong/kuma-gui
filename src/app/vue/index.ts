import { createApp } from 'vue'
import {
  createRouter,
  createWebHistory,
} from 'vue-router'

import type { ServiceDefinition } from '@/services/utils'
import { token, createInjections } from '@/services/utils'
import type { Component } from 'vue'
import type { Router, RouteRecordRaw, NavigationGuard } from 'vue-router'

type Token = ReturnType<typeof token>

type VueApp = ReturnType<typeof createApp>
export type PluginDefinition = Parameters<VueApp['use']>
export type ComponentDefinition = [string, Component]

const $ = {
  app: token<(App: Component) => Promise<VueApp>>('vue.app'),
  router: token<Router>('vue.router'),

  components: token('vue.components'),
  plugins: token<PluginDefinition[]>('vue.plugins'),
  routes: token<RouteRecordRaw[]>('vue.routes'),
  routesLabel: token<RouteRecordRaw[]>('vue.routes.label'),
  navigationGuards: token<NavigationGuard[]>('vue.routes.navigation.guards'),
}
export const services = (app: Record<string, Token>): ServiceDefinition[] => {
  return [
    [$.app, {
      service: (
        components: ComponentDefinition[],
        plugins: PluginDefinition[],
      ) => {
        return async (App: Component) => {
          const app = createApp(App)

          plugins.forEach(([...args]) => {
            app.use(...args)
          })

          components.forEach(([name, component]: [string, Component]) => {
            app.component(name, component)
          })

          return app
        }
      },
      arguments: [
        $.components,
        $.plugins,
      ],
    }],
    [$.router, {
      service: (
        env: (str: 'KUMA_BASE_PATH') => string,
        routes: RouteRecordRaw[],
        guards: NavigationGuard[],
      ) => {
        const router = createRouter({
          history: createWebHistory(env('KUMA_BASE_PATH')),
          routes,
        })

        guards.forEach((item) => {
          if (typeof item === 'function') {
            router.beforeEach(item)
          }
        })

        return router
      },
      arguments: [
        app.env,
        $.routes,
        $.navigationGuards,
      ],
    }],

    [token('vue.plugins'), {
      service: (router) => {
        return [
          [router],
        ]
      },
      arguments: [
        $.router,
      ],
      labels: [
        $.plugins,
      ],
    }],

    // TODO(jc): Label decoration. Make it so we can decorate labels
    // Temporarily turn the routesLabel label into routes for decoration
    // purposes
    [$.routes, {
      service: (routes: RouteRecordRaw[]) => routes,
      arguments: [
        $.routesLabel,
      ],
    }],
    //

    // TODO(jc): Make it so we don't need to provide a [] empty default for
    // labels
    [token('application.routes.navigation.guards'), {
      service: () => {
        return []
      },
      labels: [
        $.navigationGuards,
      ],
    }],
  ]
}

export const TOKENS = $
export const [
  useRouter,
] = createInjections(
  TOKENS.router,
)
