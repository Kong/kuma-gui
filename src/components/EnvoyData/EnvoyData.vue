<template>
  <StatusInfo
    class="envoy-data"
    :has-error="error !== undefined"
    :is-loading="isLoading"
    :error="error"
  >
    <CodeBlock
      :id="`code-block-${dataPath}`"
      language="json"
      :code="code"
      is-searchable
      :query-key="queryKey ?? `code-block-${dataPath}`"
    />
  </StatusInfo>
</template>

<script lang="ts" setup>
import { onMounted, PropType, ref, watch } from 'vue'

import Kuma from '@/services/kuma'
import CodeBlock from '@/app/common/CodeBlock.vue'
import StatusInfo from '@/components/Utils/StatusInfo.vue'

const props = defineProps({
  dataPath: {
    type: String as PropType<'xds' | 'stats' | 'clusters'>,
    required: true,
  },

  queryKey: {
    type: String,
    required: false,
    default: null,
  },

  mesh: {
    type: String,
    required: false,
    default: '',
  },

  dppName: {
    type: String,
    required: false,
    default: '',
  },

  zoneIngressName: {
    type: String,
    required: false,
    default: '',
  },

  zoneEgressName: {
    type: String,
    required: false,
    default: '',
  },
})

const isLoading = ref(true)
const error = ref<Error | undefined>(undefined)
const code = ref('')

watch(() => props.dppName, function () {
  fetchContent()
})
watch(() => props.zoneIngressName, function () {
  fetchContent()
})
watch(() => props.zoneEgressName, function () {
  fetchContent()
})

onMounted(function () {
  fetchContent()
})

async function fetchContent() {
  error.value = undefined
  isLoading.value = true

  try {
    let content = ''

    if (props.mesh !== '' && props.dppName !== '') {
      content = await Kuma.getDataplaneData({
        dataPath: props.dataPath,
        mesh: props.mesh,
        dppName: props.dppName,
      })
    } else if (props.zoneIngressName !== '') {
      content = await Kuma.getZoneIngressData({
        dataPath: props.dataPath,
        zoneIngressName: props.zoneIngressName,
      })
    } else if (props.zoneEgressName !== '') {
      content = await Kuma.getZoneEgressData({
        dataPath: props.dataPath,
        zoneEgressName: props.zoneEgressName,
      })
    }

    code.value = typeof content === 'string' ? content : JSON.stringify(content, null, 2)
  } catch (err) {
    if (err instanceof Error) {
      error.value = err
    } else {
      console.error(err)
    }
  } finally {
    isLoading.value = false
  }
}
</script>

<style lang="scss">
.code-container .k-card-body {
  position: relative;
}
</style>

<style lang="scss" scoped>
.envoy-data {
  padding: var(--spacing-md);
}

.copy-button {
  position: absolute;
  top: var(--spacing-md);
  right: var(--spacing-md);
  display: block;
}
</style>
