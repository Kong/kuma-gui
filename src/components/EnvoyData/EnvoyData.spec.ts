import { render, screen } from '@testing-library/vue'
import { rest } from 'msw'

import EnvoyData from './EnvoyData.vue'
import { server } from '@/jest-setup'

function renderComponent(props = {}) {
  return render(EnvoyData, {
    props: {
      dataPath: 'clusters',
      ...props,
    },
  })
}

describe('EnvoyData.vue', () => {
  it('renders snapshot', async () => {
    server.use(
      rest.get('http://localhost/meshes/:mesh/dataplanes/:dataplaneName/clusters', (req, res, ctx) =>
        res(ctx.status(200), ctx.json('')),
      ),
    )

    const { container } = renderComponent({
      mesh: 'foo',
      dppName: 'dataplane-test-456',
    })

    expect(container).toMatchSnapshot()
  })

  it('renders loading', () => {
    server.use(
      rest.get('http://localhost/meshes/:mesh/dataplanes/:dataplaneName/clusters', (req, res, ctx) =>
        res(ctx.status(200), ctx.json('')),
      ),
    )

    renderComponent({
      mesh: 'foo',
      dppName: 'dataplane-test-456',
    })

    expect(screen.getByTestId('status-info-loading-section')).toBeInTheDocument()
  })

  it('renders error', async () => {
    server.use(
      rest.get('http://localhost/meshes/:mesh/dataplanes/:dataplaneName/clusters', (req, res, ctx) =>
        res(ctx.status(500), ctx.json('')),
      ),
    )

    renderComponent({
      mesh: 'default',
      dppName: 'dataplane-test-456',
    })

    expect(await screen.findByText(/An error has occurred while trying to load this data./)).toBeInTheDocument()
  })
})
