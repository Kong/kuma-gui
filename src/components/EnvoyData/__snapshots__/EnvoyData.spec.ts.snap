// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EnvoyData.vue renders snapshot 1`] = `
<div>
  <div
    class="status-info"
  >
    <section
      class="empty-state-wrapper"
      data-testid="status-info-loading-section"
      data-v-5b55eb64=""
    >
      <div
        class="empty-state-title"
        data-v-5b55eb64=""
      >
        <!---->
        <div
          class="k-empty-state-title-header mt-4 mb-4"
          data-v-5b55eb64=""
        >
          
          <div
            class="card-icon mb-3"
          >
            <span
              class="kong-icon-spinner kong-icon"
              data-v-637de6fa=""
            >
              <svg
                fill="none"
                viewBox="0 0 20 20"
                xmlns="http://www.w3.org/2000/svg"
              >
                
  
                <title>
                  Loading
                </title>
                
  
                <g>
                  
    
                  <path
                    d="M5 10a5 5 0 1 0 5-5"
                    fill="none"
                    stroke="#A3BBCC"
                    stroke-linecap="round"
                    stroke-linejoin="round"
                    stroke-width="2"
                  />
                  
  
                </g>
                

              </svg>
              

            </span>
          </div>
           Data Loading... 
          
        </div>
      </div>
      <div
        class="empty-state-content"
        data-v-5b55eb64=""
      >
        <div
          class="k-empty-state-message mb-6"
          data-v-5b55eb64=""
        >
          
          
        </div>
        <div
          class="k-empty-state-cta"
          data-v-5b55eb64=""
        >
          
          <!---->
          
        </div>
      </div>
    </section>
  </div>
</div>
`;
